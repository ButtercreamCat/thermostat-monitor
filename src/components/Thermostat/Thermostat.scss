@import "../../sass/abstracts/variables";

.thermostat {
  list-style: none;
  position: relative;
  width: 360px;
  height: 360px;
  margin: 50px auto;

  &:before {
    width: 385px;
    height: 385px;
    content: "";
    border-radius: 50%;
    position: absolute;
    top: 50%;
    left: 50%;
    transform: translate(-50%, -50%);
    // border: 1px solid #5e56e8;
    background: $blue;
    display: block;
  }

  &:after {
    width: 425px;
    height: 425px;
    content: "";
    border-radius: 50%;
    position: absolute;
    top: 50%;
    left: 50%;
    transform: translate(-50%, -50%);
    border: 1px solid $blue;
    display: block;
  }

  &_line {
    width: 1px;
    height: 360px;
    background-color: $white;
    display: block;
    position: absolute;
    left: 50%;
  }

  &_circle {
    width: 290px;
    height: 290px;
    border-radius: 50%;
    background-color: $blue;
    position: absolute;
    top: 50%;
    left: 50%;
    transform: translate(-50%, -50%);
    z-index: 200;
    display: flex;
    flex-direction: column;
    align-items: center;
    justify-content: center;
    color: $white;
    font-weight: bold;
  }

  &_data {
    height: 60%;
    position: absolute;
    top: 30%;
    left: 50%;
    transform: translate((-50%, 0));
    display: flex;
    flex-direction: column;
    align-items: center;
    justify-content: space-between;
    &_temperature {
      font-size: 6.5rem;
      letter-spacing: 1.1px;
    }
    &_icon {
      font-family: "icomoon";
      line-height: 1;
      -webkit-font-smoothing: antialiased;
      -moz-osx-font-smoothing: grayscale;
      font-size: 2rem;
    }

    .icon-cool:before {
      content: "\e902";
    }

    .icon-heat:before {
      content: "\e904";
    }
    .icon-fan-on:before {
      content: "\e903";
    }

    .icon-auto:before {
      content: "\e900";
    }

    .icon-off:before {
      content: "\e906";
    }
    &_status {
      font-size: 1.8rem;
    }
  }

  &_wave {
    position: absolute;
    width: 300px;
    height: 300px;
    background: rgba($white, 0.5);
    border-radius: 45% 47% 44% 42%;
    top: 50%;
    left: 50%;
    transform: translate(-50%, -50%) rotate(0deg);
    transition: width 2s ease-out, height 2s ease-out, opacity 2s ease-out;
    animation: move 10s linear infinite;
    z-index: 100;
  }

  &_wave.off {
    width: 280px;
    height: 280px;
    opacity: 0;
  }

  &_wave:nth-child(2) {
    border-radius: 38% 46% 43% 47%;
    transform: translate(-50%, -50%) rotate(120deg);
    animation: move1 15s linear infinite;
  }

  &_wave:nth-child(3) {
    border-radius: 42% 46% 37% 40%;
    transform: translate(-50%, -50%) rotate(-120deg);
    animation: move2 13s linear infinite;
  }
}

@keyframes move {
  100% {
    transform: translate(-50%, -50%) rotate(360deg);
  }
}

@keyframes move1 {
  100% {
    transform: translate(-50%, -50%) rotate(480deg);
  }
}

@keyframes move2 {
  100% {
    transform: translate(-50%, -50%) rotate(240deg);
  }
}

@media only screen and (max-width: 667px) {
  .thermostat {
    width: 230px;
    height: 230px;
    margin: 50px auto;

    &:before {
      width: 255px;
      height: 255px;
    }

    &:after {
      width: 285px;
      height: 285px;
    }

    &_line {
      height: 230px;
    }

    &_circle {
      width: 160px;
      height: 160px;
    }

    &_data {
      &_temperature {
        font-size: 4rem;
        letter-spacing: 1.1px;
      }
      &_icon {
        font-size: 1.6rem;
      }
      &_status {
        font-size: 1.2rem;
      }
    }

    &_wave {
      width: 170px;
      height: 170px;
    }

    &_wave.off {
      width: 150px;
      height: 150px;
    }
  }
}
