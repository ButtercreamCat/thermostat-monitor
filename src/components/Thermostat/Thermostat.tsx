import React, { Component } from "react";
import "./Thermostat.scss";

interface ThermostatProps {
  status: string;
}

class Thermostat extends Component<ThermostatProps> {
  render() {
    const { status } = this.props;

    const eleArr: JSX.Element[] = [];
    let rotation = 0;
    while (rotation < 180) {
      eleArr.push(
        <div
          className="thermostat_line"
          style={{ transform: `rotate(${rotation}deg)` }}
          key={rotation}
        ></div>
      );
      rotation += 2;
    }
    return (
      <div className="thermostat">
        <div className="thermostat_circle">
          <div className="thermostat_data">
            <div className="thermostat_data_temperature">23</div>
            <div
              className={`thermostat_data_icon icon-${status.toLocaleLowerCase()}`}
            ></div>
            <div className="thermostat_data_status">{status}</div>
          </div>
        </div>
        <div
          className={`thermostat_wave ${
            status.toLocaleLowerCase() === "off" ? "off" : ""
          }`}
        ></div>
        <div
          className={`thermostat_wave ${
            status.toLocaleLowerCase() === "off" ? "off" : ""
          }`}
        ></div>
        <div
          className={`thermostat_wave ${
            status.toLocaleLowerCase() === "off" ? "off" : ""
          }`}
        ></div>
        {eleArr}
      </div>
    );
  }
}

export default Thermostat;
