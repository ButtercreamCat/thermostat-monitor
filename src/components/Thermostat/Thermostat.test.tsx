import React from "react";
import { shallow, ShallowWrapper } from "enzyme";
import Thermostat from "./Thermostat";

let wrapper: ShallowWrapper<any, Readonly<{}>, React.Component<{}, {}, any>>;

beforeEach(() => {
  wrapper = shallow(<Thermostat status="Heat"></Thermostat>);
});

test("should render Thermostat correctly", () => {
  expect(wrapper).toMatchSnapshot();
});

test("should Thermostat displays the correct status", () => {
  expect(
    wrapper
      .find(".thermostat_data_icon")
      .hasClass("thermostat_data_icon icon-heat")
  ).toBe(true);
  expect(wrapper.find(".thermostat_data_status").text()).toBe("Heat");
});

test("should Thermostat wave element has the correct class name", () => {
  wrapper = shallow(<Thermostat status="Off"></Thermostat>);
  expect(
    wrapper
      .find(".thermostat_wave")
      .at(0)
      .hasClass("thermostat_wave off")
  ).toBe(true);
});
