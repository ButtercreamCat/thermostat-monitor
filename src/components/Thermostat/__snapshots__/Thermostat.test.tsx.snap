// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Thermostat correctly 1`] = `
<div
  className="thermostat"
>
  <div
    className="thermostat_circle"
  >
    <div
      className="thermostat_data"
    >
      <div
        className="thermostat_data_temperature"
      >
        23
      </div>
      <div
        className="thermostat_data_icon icon-heat"
      />
      <div
        className="thermostat_data_status"
      >
        Heat
      </div>
    </div>
  </div>
  <div
    className="thermostat_wave "
  />
  <div
    className="thermostat_wave "
  />
  <div
    className="thermostat_wave "
  />
  <div
    className="thermostat_line"
    key="0"
    style={
      Object {
        "transform": "rotate(0deg)",
      }
    }
  />
  <div
    className="thermostat_line"
    key="2"
    style={
      Object {
        "transform": "rotate(2deg)",
      }
    }
  />
  <div
    className="thermostat_line"
    key="4"
    style={
      Object {
        "transform": "rotate(4deg)",
      }
    }
  />
  <div
    className="thermostat_line"
    key="6"
    style={
      Object {
        "transform": "rotate(6deg)",
      }
    }
  />
  <div
    className="thermostat_line"
    key="8"
    style={
      Object {
        "transform": "rotate(8deg)",
      }
    }
  />
  <div
    className="thermostat_line"
    key="10"
    style={
      Object {
        "transform": "rotate(10deg)",
      }
    }
  />
  <div
    className="thermostat_line"
    key="12"
    style={
      Object {
        "transform": "rotate(12deg)",
      }
    }
  />
  <div
    className="thermostat_line"
    key="14"
    style={
      Object {
        "transform": "rotate(14deg)",
      }
    }
  />
  <div
    className="thermostat_line"
    key="16"
    style={
      Object {
        "transform": "rotate(16deg)",
      }
    }
  />
  <div
    className="thermostat_line"
    key="18"
    style={
      Object {
        "transform": "rotate(18deg)",
      }
    }
  />
  <div
    className="thermostat_line"
    key="20"
    style={
      Object {
        "transform": "rotate(20deg)",
      }
    }
  />
  <div
    className="thermostat_line"
    key="22"
    style={
      Object {
        "transform": "rotate(22deg)",
      }
    }
  />
  <div
    className="thermostat_line"
    key="24"
    style={
      Object {
        "transform": "rotate(24deg)",
      }
    }
  />
  <div
    className="thermostat_line"
    key="26"
    style={
      Object {
        "transform": "rotate(26deg)",
      }
    }
  />
  <div
    className="thermostat_line"
    key="28"
    style={
      Object {
        "transform": "rotate(28deg)",
      }
    }
  />
  <div
    className="thermostat_line"
    key="30"
    style={
      Object {
        "transform": "rotate(30deg)",
      }
    }
  />
  <div
    className="thermostat_line"
    key="32"
    style={
      Object {
        "transform": "rotate(32deg)",
      }
    }
  />
  <div
    className="thermostat_line"
    key="34"
    style={
      Object {
        "transform": "rotate(34deg)",
      }
    }
  />
  <div
    className="thermostat_line"
    key="36"
    style={
      Object {
        "transform": "rotate(36deg)",
      }
    }
  />
  <div
    className="thermostat_line"
    key="38"
    style={
      Object {
        "transform": "rotate(38deg)",
      }
    }
  />
  <div
    className="thermostat_line"
    key="40"
    style={
      Object {
        "transform": "rotate(40deg)",
      }
    }
  />
  <div
    className="thermostat_line"
    key="42"
    style={
      Object {
        "transform": "rotate(42deg)",
      }
    }
  />
  <div
    className="thermostat_line"
    key="44"
    style={
      Object {
        "transform": "rotate(44deg)",
      }
    }
  />
  <div
    className="thermostat_line"
    key="46"
    style={
      Object {
        "transform": "rotate(46deg)",
      }
    }
  />
  <div
    className="thermostat_line"
    key="48"
    style={
      Object {
        "transform": "rotate(48deg)",
      }
    }
  />
  <div
    className="thermostat_line"
    key="50"
    style={
      Object {
        "transform": "rotate(50deg)",
      }
    }
  />
  <div
    className="thermostat_line"
    key="52"
    style={
      Object {
        "transform": "rotate(52deg)",
      }
    }
  />
  <div
    className="thermostat_line"
    key="54"
    style={
      Object {
        "transform": "rotate(54deg)",
      }
    }
  />
  <div
    className="thermostat_line"
    key="56"
    style={
      Object {
        "transform": "rotate(56deg)",
      }
    }
  />
  <div
    className="thermostat_line"
    key="58"
    style={
      Object {
        "transform": "rotate(58deg)",
      }
    }
  />
  <div
    className="thermostat_line"
    key="60"
    style={
      Object {
        "transform": "rotate(60deg)",
      }
    }
  />
  <div
    className="thermostat_line"
    key="62"
    style={
      Object {
        "transform": "rotate(62deg)",
      }
    }
  />
  <div
    className="thermostat_line"
    key="64"
    style={
      Object {
        "transform": "rotate(64deg)",
      }
    }
  />
  <div
    className="thermostat_line"
    key="66"
    style={
      Object {
        "transform": "rotate(66deg)",
      }
    }
  />
  <div
    className="thermostat_line"
    key="68"
    style={
      Object {
        "transform": "rotate(68deg)",
      }
    }
  />
  <div
    className="thermostat_line"
    key="70"
    style={
      Object {
        "transform": "rotate(70deg)",
      }
    }
  />
  <div
    className="thermostat_line"
    key="72"
    style={
      Object {
        "transform": "rotate(72deg)",
      }
    }
  />
  <div
    className="thermostat_line"
    key="74"
    style={
      Object {
        "transform": "rotate(74deg)",
      }
    }
  />
  <div
    className="thermostat_line"
    key="76"
    style={
      Object {
        "transform": "rotate(76deg)",
      }
    }
  />
  <div
    className="thermostat_line"
    key="78"
    style={
      Object {
        "transform": "rotate(78deg)",
      }
    }
  />
  <div
    className="thermostat_line"
    key="80"
    style={
      Object {
        "transform": "rotate(80deg)",
      }
    }
  />
  <div
    className="thermostat_line"
    key="82"
    style={
      Object {
        "transform": "rotate(82deg)",
      }
    }
  />
  <div
    className="thermostat_line"
    key="84"
    style={
      Object {
        "transform": "rotate(84deg)",
      }
    }
  />
  <div
    className="thermostat_line"
    key="86"
    style={
      Object {
        "transform": "rotate(86deg)",
      }
    }
  />
  <div
    className="thermostat_line"
    key="88"
    style={
      Object {
        "transform": "rotate(88deg)",
      }
    }
  />
  <div
    className="thermostat_line"
    key="90"
    style={
      Object {
        "transform": "rotate(90deg)",
      }
    }
  />
  <div
    className="thermostat_line"
    key="92"
    style={
      Object {
        "transform": "rotate(92deg)",
      }
    }
  />
  <div
    className="thermostat_line"
    key="94"
    style={
      Object {
        "transform": "rotate(94deg)",
      }
    }
  />
  <div
    className="thermostat_line"
    key="96"
    style={
      Object {
        "transform": "rotate(96deg)",
      }
    }
  />
  <div
    className="thermostat_line"
    key="98"
    style={
      Object {
        "transform": "rotate(98deg)",
      }
    }
  />
  <div
    className="thermostat_line"
    key="100"
    style={
      Object {
        "transform": "rotate(100deg)",
      }
    }
  />
  <div
    className="thermostat_line"
    key="102"
    style={
      Object {
        "transform": "rotate(102deg)",
      }
    }
  />
  <div
    className="thermostat_line"
    key="104"
    style={
      Object {
        "transform": "rotate(104deg)",
      }
    }
  />
  <div
    className="thermostat_line"
    key="106"
    style={
      Object {
        "transform": "rotate(106deg)",
      }
    }
  />
  <div
    className="thermostat_line"
    key="108"
    style={
      Object {
        "transform": "rotate(108deg)",
      }
    }
  />
  <div
    className="thermostat_line"
    key="110"
    style={
      Object {
        "transform": "rotate(110deg)",
      }
    }
  />
  <div
    className="thermostat_line"
    key="112"
    style={
      Object {
        "transform": "rotate(112deg)",
      }
    }
  />
  <div
    className="thermostat_line"
    key="114"
    style={
      Object {
        "transform": "rotate(114deg)",
      }
    }
  />
  <div
    className="thermostat_line"
    key="116"
    style={
      Object {
        "transform": "rotate(116deg)",
      }
    }
  />
  <div
    className="thermostat_line"
    key="118"
    style={
      Object {
        "transform": "rotate(118deg)",
      }
    }
  />
  <div
    className="thermostat_line"
    key="120"
    style={
      Object {
        "transform": "rotate(120deg)",
      }
    }
  />
  <div
    className="thermostat_line"
    key="122"
    style={
      Object {
        "transform": "rotate(122deg)",
      }
    }
  />
  <div
    className="thermostat_line"
    key="124"
    style={
      Object {
        "transform": "rotate(124deg)",
      }
    }
  />
  <div
    className="thermostat_line"
    key="126"
    style={
      Object {
        "transform": "rotate(126deg)",
      }
    }
  />
  <div
    className="thermostat_line"
    key="128"
    style={
      Object {
        "transform": "rotate(128deg)",
      }
    }
  />
  <div
    className="thermostat_line"
    key="130"
    style={
      Object {
        "transform": "rotate(130deg)",
      }
    }
  />
  <div
    className="thermostat_line"
    key="132"
    style={
      Object {
        "transform": "rotate(132deg)",
      }
    }
  />
  <div
    className="thermostat_line"
    key="134"
    style={
      Object {
        "transform": "rotate(134deg)",
      }
    }
  />
  <div
    className="thermostat_line"
    key="136"
    style={
      Object {
        "transform": "rotate(136deg)",
      }
    }
  />
  <div
    className="thermostat_line"
    key="138"
    style={
      Object {
        "transform": "rotate(138deg)",
      }
    }
  />
  <div
    className="thermostat_line"
    key="140"
    style={
      Object {
        "transform": "rotate(140deg)",
      }
    }
  />
  <div
    className="thermostat_line"
    key="142"
    style={
      Object {
        "transform": "rotate(142deg)",
      }
    }
  />
  <div
    className="thermostat_line"
    key="144"
    style={
      Object {
        "transform": "rotate(144deg)",
      }
    }
  />
  <div
    className="thermostat_line"
    key="146"
    style={
      Object {
        "transform": "rotate(146deg)",
      }
    }
  />
  <div
    className="thermostat_line"
    key="148"
    style={
      Object {
        "transform": "rotate(148deg)",
      }
    }
  />
  <div
    className="thermostat_line"
    key="150"
    style={
      Object {
        "transform": "rotate(150deg)",
      }
    }
  />
  <div
    className="thermostat_line"
    key="152"
    style={
      Object {
        "transform": "rotate(152deg)",
      }
    }
  />
  <div
    className="thermostat_line"
    key="154"
    style={
      Object {
        "transform": "rotate(154deg)",
      }
    }
  />
  <div
    className="thermostat_line"
    key="156"
    style={
      Object {
        "transform": "rotate(156deg)",
      }
    }
  />
  <div
    className="thermostat_line"
    key="158"
    style={
      Object {
        "transform": "rotate(158deg)",
      }
    }
  />
  <div
    className="thermostat_line"
    key="160"
    style={
      Object {
        "transform": "rotate(160deg)",
      }
    }
  />
  <div
    className="thermostat_line"
    key="162"
    style={
      Object {
        "transform": "rotate(162deg)",
      }
    }
  />
  <div
    className="thermostat_line"
    key="164"
    style={
      Object {
        "transform": "rotate(164deg)",
      }
    }
  />
  <div
    className="thermostat_line"
    key="166"
    style={
      Object {
        "transform": "rotate(166deg)",
      }
    }
  />
  <div
    className="thermostat_line"
    key="168"
    style={
      Object {
        "transform": "rotate(168deg)",
      }
    }
  />
  <div
    className="thermostat_line"
    key="170"
    style={
      Object {
        "transform": "rotate(170deg)",
      }
    }
  />
  <div
    className="thermostat_line"
    key="172"
    style={
      Object {
        "transform": "rotate(172deg)",
      }
    }
  />
  <div
    className="thermostat_line"
    key="174"
    style={
      Object {
        "transform": "rotate(174deg)",
      }
    }
  />
  <div
    className="thermostat_line"
    key="176"
    style={
      Object {
        "transform": "rotate(176deg)",
      }
    }
  />
  <div
    className="thermostat_line"
    key="178"
    style={
      Object {
        "transform": "rotate(178deg)",
      }
    }
  />
</div>
`;
